File: conplot.rb
#!/usr/bin/env ruby
#
# conplot.rb -- reads a column of numbers and plots them on the console
# Main site: http://github.com/brendano/conplot
# Introduction: http://anyall.org/blog/2008/05/conplot-a-console-plotter/
# Brendan O'Connor - brenocon@gmail.com - anyall.org


require 'pp'

module ConPlot
  class Plot
    def initialize
      @x_min = nil
      @x_max = nil
      @y_min = nil
      @y_max = nil

      @x = []
      @y = []

      @plot_left_margin = 6
      
    end
    
    def console_rows
      23
    end

    def console_cols
      80
    end

    def print_console
      plot_vert_offset = 0
      plot_horiz_offset = @plot_left_margin
      plot_width =  console_cols - plot_horiz_offset
      plot_height = console_rows - plot_vert_offset

      screen = (0...plot_height).map { [' ']*plot_width }

      raise "vectors not aligned" unless @x.size == @y.size
      (0...@x.size).each do |i|
        screen_x = (@x[i] - @x_min).to_f / (@x_max - @x_min) * plot_width + plot_horiz_offset
        screen_y = (@y[i] - @y_min).to_f / (@y_max - @y_min) * plot_height + plot_vert_offset
        screen[screen_y.to_i][screen_x.to_i] = 'o'
      end

      add_vert_ticks(screen, plot_height)
      add_horiz_ticks(screen, plot_width, plot_horiz_offset)

      screen.reverse.each do |row|
        # p row
        puts row.join("")
      end
    end

    # should special case when x's are all sequential integers or so
    def add_horiz_ticks(screen, plot_width, plot_horiz_offset)
      screen_tick_points = [ [0,:left], [plot_width-1,:right] ]
      screen_tick_points.each do |screen_tick_x, justify|
        x_range = @x_max - @x_min
        tick_point = (screen_tick_x.to_f / screen[0].size) * x_range + @x_min
        mark_str = reasonable_num_str(tick_point, x_range)
        screen_x = plot_horiz_offset + screen_tick_x
        landing_spots = (justify==:left) ? (screen_x .. (screen_x+mark_str.size)).to_a :
          ( (screen_x - mark_str.size) .. screen_x).to_a
        (0...mark_str.size).each do |i|
          screen[0][ landing_spots[i] ] = [ mark_str[i] ].pack("U*")
        end
      end
    end

    def reasonable_num_str(tick_point, range)
      s= "%f"   % tick_point if range <= 0.5
      s= "%.3f" % tick_point if range > 0.5
      s= "%.2f" % tick_point if range > 3
      s= "%.1f" % tick_point if range > 30
      s= "%.0f" % tick_point if range > 3000
      s
    end

    def add_vert_ticks(screen, plot_height)
      # tick_points = [@y_min, @y_min + (@y_max-@y_min)/2,  @y_max]
      # tick_points.each do |tick_point|
      screen_tick_points = [0, (screen.size/4).to_i, (screen.size/2).to_i, (screen.size*3/4).to_i, screen.size-1]
      screen_tick_points.each do |screen_tick_y|
        y_range = @y_max - @y_min
        tick_point = (screen_tick_y.to_f / screen.size) * y_range + @y_min
        mark_str = reasonable_num_str(tick_point, y_range)
        (0...mark_str.size).each do |i|
          screen[screen_tick_y.to_i][i] = [mark_str[i]].pack("U*")
        end
      end
    end

        # screen_tick_y = (tick_point - @y_min).to_f / (@y_max-@y_min) * (plot_height-1)

    def set_and_adjust_y(y)
      @y = y
      @x = (0...y.size).to_a
      range = (y.max - y.min).abs
      @y_min = y.min - range*0.05
      @y_max = y.max + range*0.05
      @x_min = 0
      @x_max = y.size - 1
    end

    def set_and_adjust_scatterplot( xy_pairs )
      raise "not implemented"
    end
  end
end


if $0 == __FILE__
  numbers = $stdin.readlines.map do |l| 
    if l =~ /^\s*$/
      nil
    else
      begin;  l.to_f
      rescue; nil
      end
    end
  end.compact
  if numbers.empty?
    puts "No inputs."
    exit
  end
  plot = ConPlot::Plot.new
  plot.set_and_adjust_y(numbers)
  plot.print_console
end

File: conplot.py
#!/usr/bin/env python
# designed for python2.5 i think
from __future__ import division
import sys,os


# crazy stuff to get current terminal size, from http://pdos.csail.mit.edu/~cblake/cls/cls.py
# sadly, the simple `stty -a` doesn't work because it wants to look at the
# stdin to learn the terminal device, but that's occupied

def ioctl_GWINSZ(fd):                  #### TABULATION FUNCTIONS
    try:                                ### Discover terminal width
        import fcntl, termios, struct, os
        cr = struct.unpack('hh', fcntl.ioctl(fd, termios.TIOCGWINSZ, '1234'))
    except:
        return None
    return cr

def terminal_size():                    ### decide on *some* terminal size
    cr = ioctl_GWINSZ(0) or ioctl_GWINSZ(1) or ioctl_GWINSZ(2)  # try open fds
    if not cr:                                                  # ...then ctty
        try:
            fd = os.open(os.ctermid(), os.O_RDONLY)
            cr = ioctl_GWINSZ(fd)
            os.close(fd)
        except:
            pass
    if not cr:                            # env vars or finally defaults
        try:
            cr = (env['LINES'], env['COLUMNS'])
        except:
            cr = (25, 80)
    return int(cr[1]), int(cr[0])         # reverse rows, cols

######################

class Plot:
  def __init__(_):
    _.x_min = _.x_max = _.y_min = _.y_max = None
    _.x = []
    _.y = []
    _.plot_left_margin = 6
    _.find_console_dimensions()

  def find_console_dimensions(self):
    self.console_cols, self.console_rows = (80,25)
    self.console_rows -= 3   # compensate for shell prompt room
#     self.console_cols, self.console_rows = terminal_size()
#     print (self.console_rows,self.console_cols)
#     if self.console_rows > 25:
#       self.console_rows = 25 + int(0.5*(self.console_rows - 25))

  def print_console(_):
    bot_mar = 1
    left_mar = _.plot_left_margin
    plot_width =  _.console_cols - left_mar
    plot_height = _.console_rows - bot_mar

    print (_.console_rows, _.console_cols)
    screen = [[' ']*_.console_cols for x in range(_.console_rows)]

    assert len(_.x) == len(_.y), "vectors not aligned"
    xdenom = (_.x_max - _.x_min)
    ydenom = (_.y_max - _.y_min)
    for i in range(len(_.x)):
      screen_x = (_.x[i] - _.x_min) / xdenom * (plot_width-1) + left_mar
      screen_y = (_.y[i] - _.y_min) / ydenom * (plot_height-1) + bot_mar
      #screen_y = min(screen_y, _.console_rows-1)  # not quite right
      #print screen_y,screen_x
      screen[int(round(screen_y))][int(round(screen_x))] = 'o'
    _.add_vert_ticks(screen)
    _.add_horiz_ticks(screen, plot_width, left_mar)
    for row in reversed(screen):
      print "".join(row)

  def add_vert_ticks(_,screen):
    fmt = lambda x: ("%.4g" % x) if x>=0 else ("%.3g" % x)
    screen[-1][0:5] = "%5s" % fmt(_.y_max) 
    screen[1][0:5]  = "%5s" % fmt(_.y_min)
    bot_mar = 1
    plot_height = _.console_rows - bot_mar
    for ypos in (0.25, .5, .75):
      ydenom = (_.y_max - _.y_min)
      y = _.y_min + ypos*ydenom
      screen_y = (y - _.y_min) / ydenom * (plot_height-1) + bot_mar
      screen[int(round(screen_y))][:5] = "%5s" % fmt(y)
    #screen[0][0]='_'
    #screen[0][1]='|'
  def add_horiz_ticks(self, screen, plot_width, left_mar):
    s = str(self.x_max)
    print repr(s)
    screen[0][-len(s):] = s
  def set_and_adjust_y(self, y):
    self.y = y
    self.x = range(1, len(y)+1)
    yrange = max(y) - min(y)
    self.y_min = min(y) #- .02*yrange
    self.y_max = max(y) #+ .02*yrange
    self.x_min = min(self.x)
    self.x_max = max(self.x)
  def set_and_adjust_scatterplot(self, xy_pairs):
    raise Exception("not implemented yet, but easy to do!")


if __name__=='__main__':
  import sys,os
  numbers = []
  for line in sys.stdin:
    if line.strip() == "": continue
    if len(line.split()) == 1:
      y = float(line)
      numbers.append(y)
    if len(line.split()) >= 2:
      raise Exception("single column please")
  plot = Plot()
  plot.set_and_adjust_y(numbers)
  plot.print_console()

